docs/authentication.md

```diff
@@ -0,0 +1,318 @@
+
+```jsx
+import React, { useState, useEffect, useContext, createContext } from 'react';
+import firebase from './firebase';
+
+const authContext = createContext();
+
+// You can wrap your _app.js with this provider
+export function AuthProvider({ children }) {
+  const auth = useProvideAuth();
+  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
+}
+
+// Custom React hook to access the context
+export const useAuth = () => {
+  return useContext(authContext);
+};
+
+function useProvideAuth() {
+  // Store the user in state
+  const [user, setUser] = useState(null);
+
+  const signinWithGitHub = () => {
+    return firebase
+      .auth()
+      .signInWithPopup(new firebase.auth.GithubAuthProvider())
+      .then((response) => setUser(response.user));
+  };
+
+  const signout = () => {
+    return firebase
+      .auth()
+      .signOut()
+      .then(() => setUser(false));
+  };
+
+  useEffect(() => {
+    const unsubscribe = firebase.auth().onAuthStateChanged(handleUser);
+    return () => unsubscribe();
+  }, []);
+
+  return {
+    user,
+    signinWithGitHub,
+    signout
+  };
+}
+```
+
+If you don't want to create the interface yourself, you can use [FirebaseUI](https://github.com/firebase/firebaseui-web).
+
+<details open>
+  <summary><b>Examples</b></summary>
+  <ul>
+    <li><a href="https://github.com/vercel/next.js/tree/canary/examples/with-firebase-authentication">with-firebase-authentication</a></li>
+  </ul>
+</details
+
+### Bring Your Own Database
+
+If you have an existing database with user data, you'll likely want to utilize an open-source solution that's provider agnostic.
+
+- If you need email/password log-in, use `next-iron-session`.
+- If you need to persist session data on the server, use `next-auth`.
+- If you need to support social login (Google, Facebook, etc.), use `next-auth`.
+- If you want to use [JWTs](https://jwt.io/), use `next-auth`.
+
+Both of these libraries support either authentication pattern.
+
+<details open>
+  <summary><b>Examples</b></summary>
+  <ul>
+    <li><a href="https://github.com/vercel/next.js/tree/canary/examples/with-next-auth">with-next-auth</a></li>
+    <li><a href="https://github.com/vercel/next.js/tree/canary/examples/with-iron-session">with-iron-session</a></li>
+  </ul>
+</details>
+
+### Magic (Passwordless)
+
+Magic, which uses [passwordless login](https://magic.link/), supports the static generation pattern. Similar to Firebase, a [unique identifier](https://w3c-ccg.github.io/did-primer/) has to be created on the client-side and then forwarded as a header to log-in.
+
+```jsx
+const did = await new Magic(
+  process.env.NEXT_PUBLIC_MAGIC_PUB_KEY
+).auth.loginWithMagicLink({ email: 'your@email.com' });
+
+const authRequest = await fetch('/api/login', {
+  method: 'POST',
+  headers: { Authorization: `Bearer ${did}` }
```

Comment:
Same Firebase feedback here, we should explain how Magic works with Firebase and then link to the example to learn more.

Reply:
Removed the code snippets per your other comment.